### PR TITLE
Support for persistent sessions using local and remote (LaunchIT) MariaDB

### PR DESCRIPTION
**Pull request by amarathe84**
## Description ##
This PR adds database-backed session persistence to the Flask Copilot:

- Support for persistent sessions local and remote
- Automatic checkpointing - Saves experiment state to MariaDB each time a new molecule is generated
- Manual checkpointing - "Checkpoint" dropdown button allows on-demand saves and toggling auto-save on/off
- Session restore on reconnect - Automatically restores the last session (molecules, graph layout, reasoning messages) when reopening the app
- Project/experiment tracking - Each checkpoint stores project ID, experiment ID, and full experiment metadata
- Complete state preservation - Saves tree nodes, edges, sidebar reasoning messages, metrics history, zoom/pan settings, and property optimization config
- Duplicate prevention - Prevents duplicate molecules when restoring mid-computation sessions
- Graceful resume handling - Detects interrupted computations and attempts to resume or gracefully resets

## Minor code hardening: ##

- Moved credentials to environment variables (no hard-coding)
-Add automatic connection retry in case of network connection failure or stalled session. Also added a timeout for the overall experiment
-Fix the SSL configuration (no more `fake_flag_to_enable_tls` used previously)
- Add cleanup for sync_engine on failure - Currently only async_engine is disposed in `main()`
- Unit testing (CLI) for state consistency
- Code changes for integration with LC user management

### Checkpoint Creation & Updates ###

1. Test checkpoint creation on node arrival
2. Test checkpoint updates incrementally as nodes arrive
3. Test checkpoint created when no prior state exists
4. Test checkpoint merges with existing partial state

### Checkpoint Persistence ###
 5. Test checkpoint persists correct node count after disconnect 
 6. Test checkpoint persists correct edge count after disconnect 
 7. Test serverSessionId is correctly stored in checkpoint 
 8. Test isComputing flag is set to false in checkpoint after disconnect 
 9. Test checkpoint contains all required state fields (smiles, problemType, etc.)
  10. Test checkpoint timestamp is updated on each save

### Checkpoint Restoration ###
 11. Test checkpoint restoration via /sessions/latest endpoint
 12. Test checkpoint restoration via /sessions/{id} endpoint 
 13. Test resumed computation continues from checkpoint

### Some Corner Case Handling ###
 14. Test checkpoint handles malformed node data gracefully
  15. Test checkpoint with maximum depth computation (compute complete and restore)

### CODE REVIEW

File: db_backend/database/engine.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+# Allow a local SQLite fallback so the app can run even if MariaDB is unreachable.
+USE_SQLITE_FALLBACK = os.getenv("USE_SQLITE_FALLBACK", "0") == "1"
+
+if USE_SQLITE_FALLBACK:
+    SQLITE_PATH = os.getenv("SQLITE_PATH", "sqlite:///flaskcopilot.db")
+    DATABASE_URL = SQLITE_PATH.replace("sqlite:///", "sqlite+pysqlite:///")
+    ASYNC_DATABASE_URL = SQLITE_PATH.replace("sqlite:///", "sqlite+aiosqlite:///")
+else:
+    # Remote LLNL LaunchIT MariaDB configuration (via SSH tunnel)
+    # SSH tunnel command: ssh -L 32636:cz-marathe1-mymariadb1.apps.czapps.llnl.gov:32636 <user>@oslic.llnl.gov
```
Comment (tbennun) [inline db_backend/database/engine.py]: remove these two lines please

File: db_backend/database/schemas.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    custom_property_name: Optional[str] = Field(None, alias="customPropertyName")
+    custom_property_desc: Optional[str] = Field(None, alias="customPropertyDesc")
+    custom_property_ascending: Optional[bool] = Field(None, alias="customPropertyAscending")
+
+    # Complex nested data
+    tree_nodes: Optional[Any] = Field(None, alias="treeNodes")
+    edges: Optional[Any] = None
+    metrics_history: Optional[Any] = Field(None, alias="metricsHistory")
+    visible_metrics: Optional[Any] = Field(None, alias="visibleMetrics")
+    graph_state: Optional[Any] = Field(None, alias="graphState")
```
Comment (tbennun) [inline db_backend/database/schemas.py]: are those JSONs really `Any`, or are they `dict[str, Any]`? (or even `dict[str, str | float | int]` depending on the type)

File: db_backend/routers/projects.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+
+def generate_id(prefix: str) -> str:
+    """Generate unique ID"""
+    timestamp = int(time.time() * 1000)
+    random_str = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=9))
+    return f"{prefix}_{timestamp}_{random_str}"
+
+
+# Pydantic models
+class ExperimentData(BaseModel):
```
Comment (tbennun) [inline db_backend/routers/projects.py]: this repeats what we have in database/models, but looks better? Consolidate the two classes please

File: .gitignore [hunk 4]
```diff
@@ -149,3 +156,17 @@ __pycache__/
...
+# Playwright
+/test-results/
+/playwright-report/
+/blob-report/
+/playwright/.cache/
+/playwright/.auth/
+
+# envfiles
+.env-launchit
+.env-local
```
Comment (tbennun) [inline .gitignore]: covered by above .env-*

**Review by tbennun — CHANGES_REQUESTED** [review]

File: db_backend/database/engine.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+if not USE_SQLITE_FALLBACK:
+    SSL_CONTEXT = ssl.create_default_context()
+    SSL_CONTEXT.check_hostname = False
+    SSL_CONTEXT.verify_mode = ssl.CERT_NONE
+    SSL_KW = {"ssl": SSL_CONTEXT}
+else:
+    SSL_KW = {}
+
+# Create sync engine (SSL required by server; skip cert verification because tunnel is trusted)
+try:
```
[the remark this answers is not in the document]
Comment (amarathe84) [reply db_backend/database/engine.py]: To address this, engine creation is now lazy and happens with get_async_engine() rather than at module import time. The module-level globals are only placeholders and are not constructed engine objects. Let me know if this looks good

[the remark this answers is not in the document]
Comment (amarathe84) [reply db_backend/database/engine.py]: These were due to incorrect force-push from another branch and conflict resolution which didn't go as planned. Both of these (local cache instead of global context and refactor) have been handled now.
